sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java

```diff
@@ -137,57 +132,90 @@ public <NewInputT> FlatMapElements<NewInputT, OutputT> via(
   @Experimental(Kind.CONTEXTFUL)
   public <NewInputT> FlatMapElements<NewInputT, OutputT> via(
       Contextful<Fn<NewInputT, Iterable<OutputT>>> fn) {
-    return new FlatMapElements<>(
-        fn, fn.getClosure(), TypeDescriptors.inputOf(fn.getClosure()), outputType);
+    return new FlatMapElements<>(fn, TypeDescriptors.inputOf(fn.getClosure()), outputType);
   }
 
   @Override
   public PCollection<OutputT> expand(PCollection<? extends InputT> input) {
     checkArgument(fn != null, ".via() is required");
-    return input.apply(
-        "FlatMap",
-        ParDo.of(
-                new DoFn<InputT, OutputT>() {
-                  @ProcessElement
-                  public void processElement(ProcessContext c) throws Exception {
-                    Iterable<OutputT> res =
-                        fn.getClosure().apply(c.element(), Fn.Context.wrapProcessContext(c));
-                    for (OutputT output : res) {
-                      c.output(output);
+    if (fn instanceof Contextful) {
```

Comment:
Do I understand correctly that doing this branch every time is cheaper than the indirection of an "empty" Contextful? 

Would it make sense to create a subclass based on the Fn type rather than branch here (and below).

Reply:
Yes, half the cost of these trivial functions was going to creating and destroying a Contextful that wraps a ProcessBundleDescriptor that is passed in.

Also note that the branch only happens during pipeline construction so it won't show up on the execution side.

Reply:
Also, I didn't want to wrap with another PTransform that delineated between these two options